src/simulation.hpp

```diff
@@ -2795,11 +2871,36 @@ template <typename problem_t> void AMRSimulation<problem_t>::ReadCheckpointFile(
 		TracerPC = std::make_unique<amrex::AmrTracerParticleContainer>(this);
 		TracerPC->Restart(restart_chkfile, "tracer_particles");
 	}
+
+	// Initialize and register particle containers from checkpoint file
 	if (do_cic_particles != 0) {
 		AMREX_ASSERT(CICParticles == nullptr);
 		CICParticles = std::make_unique<quokka::CICParticleContainer>(this);
+		auto cicParticleDesc = std::make_unique<quokka::CICParticleDescriptor<problem_t>>(0, -1, -1, false);
+		cicParticleDesc->setParticleContainer(CICParticles.get());
+		particleRegister_->registerParticleType("CIC_particles", std::move(cicParticleDesc));
 		CICParticles->Restart(restart_chkfile, "CIC_particles");
 	}
+
+	if (do_rad_particles != 0) {
+		AMREX_ASSERT(RadParticles == nullptr);
+		RadParticles = std::make_unique<quokka::RadParticleContainer<problem_t>>(this);
+		auto radParticleDesc =
+		    std::make_unique<quokka::RadParticleDescriptor<problem_t>>(-1, quokka::RadParticleLumIdx, quokka::RadParticleBirthTimeIdx, false);
```

Comment:
I don't think it's necessary to create a `unique_ptr<quokka::RadParticleContainer>` You can just create a `RadParticleContainer` and then `std::move` it to the particleRegister_.

Reply:
Actually it is necessary. This is because particleRegister_ is a unique_ptr, and in registerParticleType, particleRegister_ 'emplace' std::move(descriptor), which has to be a unique_ptr.

Reply:
This does the automatic memory management as expected, and std::move transfers ownership of the descriptor to the registry.

Reply:
You can `std::move` any object:
```
#include <iomanip>
#include <iostream>
#include <string>
#include <utility>
#include <vector>
 
int main()
{
    std::string str = "Salut";
    std::vector<std::string> v;
 
    // uses the push_back(const T&) overload, which means
    // we'll incur the cost of copying str
    v.push_back(str);
    std::cout << "After copy, str is " << std::quoted(str) << '\n';
 
    // uses the rvalue reference push_back(T&&) overload,
    // which means no strings will be copied; instead, the contents
    // of str will be moved into the vector. This is less
    // expensive, but also means str might now be empty.
    v.push_back(std::move(str));
    std::cout << "After move, str is " << std::quoted(str) << '\n';
 
    std::cout << "The contents of the vector are {" << std::quoted(v[0])
              << ", " << std::quoted(v[1]) << "}\n";
}
```

Reply:
Ok, I see now why you did this. This makes sense if we have `AMRSimulation<problem_t>::CICParticles` as a member variable. But if it's a unique_ptr, that means we can't access it as a member variable once it's been moved. So it should be either a `shared_ptr`, or we should just get a weak pointer to it from the PhysicsParticleRegister whenever we need to access it directly.